src/librustc/lint.rs

```diff
@@ -191,134 +191,145 @@ impl<'a> LintDiagnosticBuilder<'a> {
     }
 }
 
-pub fn struct_lint_level<'s>(
+pub fn struct_lint_level<'s, 'd>(
     sess: &'s Session,
     lint: &'static Lint,
     level: Level,
     src: LintSource,
     span: Option<MultiSpan>,
-    decorate: impl for<'a> FnOnce(LintDiagnosticBuilder<'a>)) {
-
-    // FIXME: Move the guts of this function into a fn which takes dyn Fn to reduce code bloat.
-    let mut err = match (level, span) {
-        (Level::Allow, _) => { return; },
-        (Level::Warn, Some(span)) => sess.struct_span_warn(span, ""),
-        (Level::Warn, None) => sess.struct_warn(""),
-        (Level::Deny, Some(span)) | (Level::Forbid, Some(span)) => sess.struct_span_err(span, ""),
-        (Level::Deny, None) | (Level::Forbid, None) => sess.struct_err(""),
-    };
+    decorate: impl for<'a> FnOnce(LintDiagnosticBuilder<'a>) + 'd,
+) {
+    // Avoid codegen bloat from monomorphization due to taking decorate as an impl parameter.
+    fn struct_lint_level_impl(
+        sess: &'s Session,
+        lint: &'static Lint,
+        level: Level,
+        src: LintSource,
+        span: Option<MultiSpan>,
+        decorate: Box<dyn for<'b> FnOnce(LintDiagnosticBuilder<'b>) + 'd>) {
```

Comment:
Why not `&dyn` here instead?

Reply:
`decorate` is FnOnce, so I can't call it through `&dyn`  [which matters in cases like these, where a captured variable (in this case, overlap) is moved out of its environment](https://github.com/jumbatm/rust/blob/b0f6d3d368a6c43548b445fdac1c61fe5235967d/src/librustc/traits/specialize/mod.rs#L333). Using `FnOnce` also avoids callers having rewrite their `decorate` function to be `Fn` when we'll only call it once anyway, giving them the most flexibility in terms of what kind of captures the closure can do.

I'll push changes that remove the Box anyway, to see if we get any sizeable performance gain. I was hoping the time for allocation of the box would be small enough in comparison to the previous time we were spending formatting, but given the results we're seeing, it couldn't hurt to give this a try.

Reply:
Ah right; we might want to try the other changes first though before doing the unboxing to see what the full impact of unboxing would be.